File: truss-compiler/src/main/java/truss/compiler/printing/TextPrinter.java
package truss.compiler.printing;

import truss.compiler.syntax.LiteralType;
import org.apache.commons.lang.StringUtils;
import org.apache.commons.lang.Validate;

import java.io.Writer;

public class TextPrinter implements Printer {
    private final Writer writer;
    private int indent;
    private String indentation;

    public TextPrinter(Writer writer) {
        Validate.notNull(writer, "writer");

        this.writer = writer;
        indentation = "    ";
    }

    public String getIndentation() {
        return indentation;
    }

    public void setIndentation(String indentation) {
        this.indentation = indentation;
    }

    @Override
    public void identifier(String identifier) throws Exception {
        writer.write(identifier);
    }

    @Override
    public void indent() throws Exception {
        indent++;
    }

    @Override
    public void keyword(String keyword) throws Exception {
        writer.write(keyword);
    }

    @Override
    public void lead() throws Exception {
        writer.write(StringUtils.repeat(indentation, indent));
    }

    @Override
    public void literal(String value, LiteralType type) throws Exception {
        writer.write(value);
    }

    @Override
    public void nl() throws Exception {
        writer.write("\n");
    }

    @Override
    public void ws() throws Exception {
        writer.write(" ");
    }

    @Override
    public void syntax(String syntax) throws Exception {
        writer.write(syntax);
    }

    @Override
    public void unIndent() throws Exception {
        indent--;
    }
}

File: Truss.Compiler/Printing/TextPrinter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Printing {
    public class TextPrinter : IPrinter {
        private readonly TextWriter _writer;
        private int _indent;

        public TextPrinter(TextWriter writer) {
            if (writer == null) {
                throw new ArgumentNullException("writer");
            }

            _writer = writer;
            Indentation = "    ";
        }

        public string Indentation { get; set; }

        public void Identifier(string identifier) {
            _writer.Write(identifier);
        }

        public void Indent() {
            _indent++;
        }

        public void Keyword(string keyword) {
            _writer.Write(keyword);
        }

        public void Lead() {
            for (int i = 0; i < _indent; i++) {
                _writer.Write(Indentation);
            }
        }

        public void Literal(string value, LiteralType type) {
            _writer.Write(value);
        }

        public void Nl() {
            _writer.Write(Environment.NewLine);
        }

        public void Ws() {
            _writer.Write(" ");
        }

        public void Syntax(string syntax) {
            _writer.Write(syntax);
        }

        public void UnIndent() {
            _indent--;
        }
    }
}
